### ISSUE TITLE
Cross reproduction of scripts across machines

### ISSUE THREAD

**Issue by sanket1729** [apoelstra/local-nix-ci#1]
Following are the things that I had to change in order to make the rust-miniscript nix tests on my machine.

1) The directory structure that this script assumes is somewhat different as documented in the README. I had to change the following in the `generate-repo-json.sh` to make it work with the setup where we do not assume there are directories for each branch. 
 
```diff
-DATADIR=$(realpath "$GITDIR/../../")
+DATADIR=$(realpath "$GITDIR/../")
```

2) The README `origin` points to "main" repository. I believe that this is more common nomencleture, we refer to "origin" as out fork of the repo and "upstream" to refer to main repo.

```diff
-       "gitUrl": "$(git remote get-url origin)",
+       "gitUrl": "$(git remote get-url upstream)"
```

3) The files for bitcoin.nix and default.nix are not present in the repo and have hard coded paths. They also assume a directory structure with a version. I modified this by setting the `url` in `fetchGit` to the local installation of my bitcoin source code. 
I think the ref field here could be modified to point to `24.x` branch etc to build the correct binary. Obviously, this also assumes that you have bitcoin downloaded in the local tree.

4) I did not utility tools like `banner` and `nom` (NixOs Monitor) installed on my machine which is required by the script.

5) The script also has hard-coded paths for `OUT_DIR` variable. This assumes that the current user has write permissions on `/nix/var/nix/gcroots/per-user/${USER}/ci-output`. The `ci-output` folder must also be created.

6) Lastly, nix-env by default takes a lot of time to execute this script as it uses only 1 core by default. This can be changed by setting `max-jobs` in nix.conf (/etc/nix/nix.conf).

**Comment by sanket1729** [comment]
Another question:

Why all the symlink stuff across all repos? Would it not be simple to pass a parameter telling
which script to build. All these are linked to `check-pr.nix`. So, the user is expected to update this link
everytime they run?

**Comment by sanket1729** [comment]
There also seems to be some maintenance in these scripts everytime we add/remove a feature.

**Comment by apoelstra** [comment]
> Would it not be simple to pass a parameter telling which script to build.

Then they'd need to specify a nixfile every time instead of just running `../test-pr.sh <number>` regardless of repo.

> All these are linked to `check-pr.nix`. So, the user is expected to update this link everytime they run?

I don't understand. What link do they need to update?

**Comment by apoelstra** [comment]
> There also seems to be some maintenance in these scripts everytime we add/remove a feature.

Yes. We could maybe extract a feature list automatically from the Cargo.toml. I would be interested in doing that. It would be neat if we could produce a random test matrix which targeted a particular size.
